tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -5883,163 +5888,164 @@ static bool ScanConstInitList(CodeGenModule &CGM, InitListExpr *E,
   return true;
 }
 
-static Constant *BuildConstInitializer(QualType Type, unsigned &offset,
-                                       SmallVector<Constant *, 4> &EltValList,
-                                       CodeGenTypes &Types,
-                                       bool bDefaultRowMajor);
-
-static Constant *BuildConstVector(llvm::VectorType *VT, unsigned &offset,
-                                  SmallVector<Constant *, 4> &EltValList,
-                                  QualType Type, CodeGenTypes &Types) {
-  SmallVector<Constant *, 4> Elts;
-  QualType EltTy = hlsl::GetHLSLVecElementType(Type);
-  for (unsigned i = 0; i < VT->getNumElements(); i++) {
-    Elts.emplace_back(BuildConstInitializer(EltTy, offset, EltValList, Types,
-                                            // Vector don't need major.
-                                            /*bDefaultRowMajor*/ false));
-  }
-  return llvm::ConstantVector::get(Elts);
-}
-
-static Constant *BuildConstMatrix(llvm::Type *Ty, unsigned &offset,
-                                  SmallVector<Constant *, 4> &EltValList,
-                                  QualType Type, CodeGenTypes &Types,
-                                  bool bDefaultRowMajor) {
-  QualType EltTy = hlsl::GetHLSLMatElementType(Type);
-  unsigned col, row;
-  HLMatrixLower::GetMatrixInfo(Ty, col, row);
-  llvm::ArrayType *AT = cast<llvm::ArrayType>(Ty->getStructElementType(0));
+static Constant *BuildConstInitializer(CodeGenTypes &Types, bool bDefaultRowMajor,
```

Comment:
This and accompanying method are also partially rewritten to consume the QualTypes and deal with register/memory representations, plus the usual cleanups.